CrossCompatibility/CrossCompatibility.psm1

```diff
@@ -0,0 +1,1744 @@
+
+    return $compatibilityData
+}
+
+<#
+.SYNOPSIS
+Gets all assemblies publicly available in
+the current PowerShell session.
+Skips assemblies from user modules by default.
+
+.PARAMETER All
+Include
+#>
+function Get-AvailableTypes
+{
+    param(
+        [Parameter()]
+        [switch]
+        $All
+    )
+
+    # In PS Core, we need to explicitly force the loading of all assemblies (which normally lazy-load)
+    if ($PSEdition -eq 'Core')
+    {
+        Get-ChildItem $PSHOME -Filter '*.dll' | ForEach-Object { try { Add-Type -Path $_ } catch { } }
+    }
+
+    $asms = New-Object 'System.Collections.Generic.List[System.Reflection.Assembly]'
+
+    $asmPaths = $PSHOME, (Split-Path $script:WinPSHomeModulePath)
+
+    foreach ($asm in [System.AppDomain]::CurrentDomain.GetAssemblies())
+    {
+        if ($asm.IsDynamic -or -not $asm.Location)
+        {
+            continue
+        }
+
+        if ($All -or $asm.GlobalAssemblyCache -or (Test-HasAnyPrefix $asm.Location -Prefix $asmPaths -IgnoreCase:$script:IsWindows))
+        {
+            $asms.Add($asm)
+        }
+    }
+
+    return $asms
+}
+
+<#
+.SYNOPSIS
+Get the type accelerators in the current PowerShell session.
+
+.DESCRIPTION
+Builds a dictionary of all the type accelerators defined in the current PowerShell session.
+#>
+function Get-TypeAccelerators
+{
+    $typeAccelerators = [psobject].Assembly.GetType("System.Management.Automation.TypeAccelerators")::Get.GetEnumerator()
+
+    $taTable = New-Object 'System.Collections.Generic.Dictionary[string, type]' ([System.StringComparer]::OrdinalIgnoreCase)
+
+    foreach ($taKvp in $typeAccelerators)
+    {
+        $taTable[$taKvp.Key] = $taKvp.Value
+    }
+
+    return $taTable
+}
+
+<#
+.SYNOPSIS
+Get module data about core PowerShell cmdlets.
+
+.DESCRIPTION
+Get module data about the Microsoft.PowerShell.Core pseudomodule.
+#>
+function Get-CoreModuleData
+{
+    $coreCommands = Get-Command -Module 'Microsoft.PowerShell.Core'
+
+    $coreVariables = Get-Variable | Where-Object { -not $_.Module } | ForEach-Object { $_.Name }
+    $coreAliases = Get-Alias | Where-Object { -not $_.Module } | New-AliasData
+    $coreFunctions = $coreCommands | Where-Object { $_.CommandType -eq 'Function' } | New-FunctionData
+    $coreCmdlets = $coreCommands | Where-Object { $_.CommandType -eq 'Cmdlet' } | New-CmdletData
+
+    $coreModuleData = @{}
+
+    if ($coreVariables)
+    {
+        $coreModuleData['Variables'] = $coreVariables
+    }
+
+    if ($coreAliases -and $coreAliases.get_Count() -gt 0)
+    {
+        $coreModuleData['Aliases'] = $coreAliases
+    }
+
+    if ($coreFunctions -and $coreFunctions.get_Count() -gt 0)
+    {
+        $coreModuleData['Functions'] = $coreFunctions
+    }
+
+    if ($coreCmdlets -and $coreCmdlets.get_Count() -gt 0)
+    {
+        $coreModuleData['Cmdlets'] = $coreCmdlets
+    }
+
+    return [Microsoft.PowerShell.CrossCompatibility.Data.Modules.ModuleData]$coreModuleData

```

Comment:
The hashtable coerces from setters in the object, which serialises differently depending on whether a field is `null` or an empty collection. I've been resisting writing a Hashtable-based constructor to avoid complicated constructor logic.

Maybe the functions here to process the data creation should be rewritten? I could use some education on best practices for advanced functions that aggregate data like this.